src/GHC/Num/Compat.hs

```diff
@@ -0,0 +1,149 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE UnboxedSums #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+-- |
+-- Module      :  GHC.Num.Compat
+-- Description :  Defines numeric compatibility shims that work with both
+--                ghc-bignum (GHC 9.0+) and integer-gmp (older GHCs).
+-- Copyright   :  (c) 2021 Galois, Inc.
+-- License     :  BSD3
+-- Maintainer  :  cryptol@galois.com
+-- Stability   :  provisional
+-- Portability :  portable
+module GHC.Num.Compat
+  ( -- * BigNat#
+    BigNat#
+  , bigNatAdd
+  , bigNatIsOne
+  , bigNatIsZero
+  , bigNatMul
+  , bigNatRem
+  , bigNatSqr
+  , bigNatSub
+  , bigNatSubUnsafe
+  , oneBigNat
+  , recipModBigNat
+  , shiftLBigNat
+  , shiftRBigNat
+  , testBitBigNat
+  , zeroBigNat
+
+    -- * Integer
+  , Integer(IS, IP, IN)
+  , integerRecipMod
+
+    -- * Conversions
+  , bigNatToInteger
+  , integerToBigNat
+  ) where
+
+#if defined(MIN_VERSION_ghc_bignum)
+import           GHC.Num.BigNat (BigNat#, bigNatAdd, bigNatIsOne, bigNatIsZero, bigNatMul, bigNatRem, bigNatSqr, bigNatSub, bigNatSubUnsafe)
+import qualified GHC.Num.Backend as BN
+import qualified GHC.Num.BigNat as BN
+import           GHC.Num.Integer (Integer(IS, IP, IN))
+import qualified GHC.Num.Integer as Integer
+import           GHC.Exts
+
+-- | Coerce a @BigNat#@ to an integer value.
+bigNatToInteger :: BigNat# -> Integer
+bigNatToInteger = Integer.integerFromBigNat#
+
+integerRecipMod :: Integer -> Integer -> Maybe Integer
+integerRecipMod x y =
+  case Integer.integerRecipMod# x (Integer.integerToNaturalClamp y) of
```

Comment:
Now that I look closer at this, I think that the use of [`integerToNaturalClamp`](https://hackage.haskell.org/package/ghc-bignum-1.2/docs/GHC-Num-Integer.html#v:integerToNaturalClamp) makes the behavior of `integerRecipMod` subtly different depending on whether `ghc-bignum` is used or not. When `ghc-bignum` is not used, we have:

```
λ> integerRecipMod (-1) (-2)
Just 1
```

But with `ghc-bignum`, we have:

```
λ> integerRecipMod (-1) (-2)
Nothing
```

This is because `integerToNaturalClamp` will return `0` if given a negative number, so with `ghc-bignum`, `integerRecipMod (-1) (-2)` is equivalent to `integerRecipMod (-1) 0`.

Perhaps we should use [`integerToNatural`](https://hackage.haskell.org/package/ghc-bignum-1.2/docs/GHC-Num-Integer.html#v:integerToNatural) instead of `integerToNaturalClamp`? The former will return the absolute value if the argument is negative, and that seems to mirror the pre-`ghc-bignum` behavior more closely.

Reply:
This only affects the second (modulus) argument, right? I think we can add (and satisfy) the precondition that the modulus is always strictly positive.